numba/tests/test_np_functions.py

```diff
@@ -2148,6 +2184,73 @@ def test_interp_exceptions(self):
 
         self.assertIn(complex_dtype_msg, str(e.exception))
 
+    def test_interp_non_finite_calibration(self):
+        # examples from
+        # https://github.com/numpy/numpy/issues/12951
+        pyfunc = interp
+        cfunc = jit(nopython=True)(pyfunc)
+        _check = partial(self._check_output, pyfunc, cfunc)
+
+        xp = np.array([0, 1, 9, 10])
+        fp = np.array([-np.inf, 0.1, 0.9, np.inf])
+        x = np.array([0.2, 9.5])
+        params = {'x': x, 'xp': xp, 'fp': fp}
+        _check(params)
+
+        xp = np.array([-np.inf, 1, 9, np.inf])
+        fp = np.array([0, 0.1, 0.9, 1])
+        x = np.array([0.2, 9.5])
+        params = {'x': x, 'xp': xp, 'fp': fp}
+        _check(params)
+
+    def test_interp_supplemental_tests(self):
+        # inspired by class TestInterp
+        # https://github.com/numpy/numpy/blob/f5b6850f231/numpy/lib/tests/test_function_base.py
+        pyfunc = interp
+        cfunc = jit(nopython=True)(pyfunc)
+        _check = partial(self._check_output, pyfunc, cfunc)
+
+        for size in range(1, 10):
+            xp = np.arange(size, dtype=np.double)
+            yp = np.ones(size, dtype=np.double)
+            incpts = np.array([-1, 0, size - 1, size], dtype=np.double)
+            decpts = incpts[::-1]
+
+            incres = cfunc(incpts, xp, yp)
+            decres = cfunc(decpts, xp, yp)
+            inctgt = np.array([1, 1, 1, 1], dtype=float)
+            dectgt = inctgt[::-1]
+            np.testing.assert_almost_equal(incres, inctgt)
+            np.testing.assert_almost_equal(decres, dectgt)
+
+        x = np.linspace(0, 1, 5)
+        y = np.linspace(0, 1, 5)
+        x0 = 0
+        np.testing.assert_almost_equal(cfunc(x0, x, y), x0)
+        x0 = 0.3
+        np.testing.assert_almost_equal(cfunc(x0, x, y), x0)
+        x0 = np.float32(0.3)
+        np.testing.assert_almost_equal(cfunc(x0, x, y), x0)
+        x0 = np.float64(0.3)
+        np.testing.assert_almost_equal(cfunc(x0, x, y), x0)
+        x0 = np.nan
+        np.testing.assert_almost_equal(cfunc(x0, x, y), x0)
+
+        x = np.linspace(0, 1, 5)
+        y = np.linspace(0, 1, 5) + (1 + np.linspace(0, 1, 5)) * 1.0j
+        x0 = 0.3
+        y0 = x0 + (1 + x0) * 1.0j
+        np.testing.assert_almost_equal(cfunc(x0, x, y), y0)
+
+        x = np.linspace(0, 1, 5)
+        y = np.linspace(0, 1, 5)
+        x0 = np.array(0.3)
+        np.testing.assert_almost_equal(cfunc(x0, x, y), x0)
+
+        xp = np.arange(0, 10, 0.0001)
+        fp = np.sin(xp)
+        np.testing.assert_almost_equal(cfunc(np.pi, xp, fp), 0.0)
```

Comment:
Perhaps check e.g. `pi/4` intervals in a defined out of order manner so as to jump around on the curve, this deliberately stressing the issue existing in the code without this patch?

Reply:
Done, good call